action/lib/deduplicate.js

```diff
@@ -46,6 +48,8 @@ export default async function ({ octokit, workflow_id, run_id }) {
   core.debug(inspect(prime))
 
   for (const run of cancellable) {
+    if (run.status === 'completed') continue
```

Comment:
I suppose this would work ...

but, wouldn't it better to first check if there is a completed run FIRST (before getting this deep) and if one is already completed, exit early?

Reply:
I went for the least amount of logic changes that would do the trick, my test version looked like this (sort and map no longer needed, just filter the runs and if the list is not empty just cancel the current workflow - it proved very reliable in my case, but I didn't want to force my luck :) ):

```javascript
    // filter
    const current_runs = workflow_runs
      // exclude this one
      .filter(run => run.id !== run_id)
      // filter to only runs for the same commit
      .filter(run => run.head_sha === sha)
      // filter out unsuccessful completed runs (cancelled / failed)
      .filter(run => (run.status !== 'completed') || (run.conclusion === 'success'))
  
    core.info(`found ${current_runs.length} existing runs of workflow ${workflow_id} for sha ${sha}`)
  
    if(current_runs.length > 0) {
      core.info('successful or in-progress runs found, bailing out')
      await octokit.request('POST /repos/{owner}/{repo}/actions/runs/{run_id}/cancel', {
        ...github.context.repo,
        run_id: run_id
      })
    }
```